File: fibonacci-dotnet/src/Fibonacci.Api/Algorithms/FastDoublingAlgorithm.cs
using System.Numerics;
using Fibonacci.Api.Constants;

namespace Fibonacci.Api.Algorithms;

/// <summary>
/// Fast-doubling Fibonacci — O(log n) time, O(log n) space (recursive stack).
/// Mirrors Python's sympy.fibonacci() using Lucas sequence identities:
///   F(2k)   = F(k) * [2·F(k+1) − F(k)]
///   F(2k+1) = F(k)² + F(k+1)²
/// Each recursive call halves n, yielding logarithmic depth.
/// </summary>
public sealed class FastDoublingAlgorithm : IFibonacciAlgorithm
{
    public string Name => "fast-doubling";
    public string Label => "Fast Doubling";
    public string TimeComplexity => "O(log n)";
    public string SpaceComplexity => "O(log n)";
    public int MaxN => FibonacciConstants.MaxN;
    public bool RequiresExplicitInclusion => false;

    public BigInteger Compute(int n)
    {
        ArgumentOutOfRangeException.ThrowIfNegative(n);
        return FastDouble(n).F;
    }

    private static (BigInteger F, BigInteger FNext) FastDouble(int n)
    {
        if (n == 0)
            return (BigInteger.Zero, BigInteger.One);

        var (f, fNext) = FastDouble(n >> 1);
        var c = f * (2 * fNext - f);
        var d = f * f + fNext * fNext;

        return (n & 1) == 0 ? (c, d) : (d, c + d);
    }
}

File: fibonacci-java/src/main/java/com/bretzelcoder/fibonacci/algorithm/FastDoublingAlgorithm.java
package com.bretzelcoder.fibonacci.algorithm;

import com.bretzelcoder.fibonacci.config.FibonacciConstants;
import java.math.BigInteger;
import org.springframework.core.annotation.Order;
import org.springframework.stereotype.Component;

/**
 * Fast doubling — O(log n) time, O(log n) stack.
 * The JVM counterpart of Python's {@code sympy.fibonacci()}, based on the Lucas identities:
 * <pre>
 *   F(2k)   = F(k) * [2*F(k+1) - F(k)]
 *   F(2k+1) = F(k)^2 + F(k+1)^2
 * </pre>
 * Each recursive step halves n, so the depth is logarithmic and needs no special stack.
 */
@Component
@Order(4)
public class FastDoublingAlgorithm implements FibonacciAlgorithm {

    @Override
    public String name() {
        return "fast-doubling";
    }

    @Override
    public String label() {
        return "Fast Doubling";
    }

    @Override
    public String timeComplexity() {
        return "O(log n)";
    }

    @Override
    public String spaceComplexity() {
        return "O(log n)";
    }

    @Override
    public int maxN() {
        return FibonacciConstants.MAX_N;
    }

    @Override
    public boolean requiresExplicitInclusion() {
        return false;
    }

    @Override
    public BigInteger compute(int n) {
        requireNonNegative(n);
        return fastDouble(n).current();
    }

    /** F(k) and F(k+1) computed together — the pair is what makes the doubling step possible. */
    private record Pair(BigInteger current, BigInteger next) {
    }

    private static Pair fastDouble(int n) {
        if (n == 0) {
            return new Pair(BigInteger.ZERO, BigInteger.ONE);
        }

        Pair half = fastDouble(n >> 1);
        BigInteger f = half.current();
        BigInteger fNext = half.next();

        BigInteger c = f.multiply(fNext.shiftLeft(1).subtract(f));
        BigInteger d = f.multiply(f).add(fNext.multiply(fNext));

        return (n & 1) == 0 ? new Pair(c, d) : new Pair(d, c.add(d));
    }
}
